src/blob_gc_job.cc

```diff
@@ -582,69 +582,80 @@ bool BlobGCJob::IsShutingDown() {
 }
 
 Status BlobGCJob::DigHole() {
-  Status s;
-  const auto& inputs = blob_gc_->sampled_inputs();//TODO modify to fs_sampled_inputs
+  const auto &inputs = blob_gc_->sampled_inputs();//TODO(@lhy1024) modify to fs_sampled_inputs
   assert(!inputs.empty());
-  for (std::size_t i = 0; i < inputs.size(); ++i) {
-    std::unique_ptr<PosixRandomRWFile> file;
-    // TODO(@DorianZheng) set read ahead size
-    s = OpenBlobFile(inputs[i]->file_number(), 0, db_options_, env_options_,
-                     env_, &file);
+  for (BlobFileMeta *input : inputs) {
+    Status s = DigHoleFile(input);
     if (!s.ok()) {
-      break;
+      return s;
     }
+  }
+  return Status::OK();
+}
 
-    auto record_iter = new BlobFileIterator(
-        std::move(file), inputs[i]->file_number(), inputs[i]->file_size(),
-        blob_gc_->titan_cf_options());
-
-    record_iter->SeekToFirst();
-    assert(record_iter->Valid());
-    const auto& first_blob_handle = record_iter->GetBlobIndex().blob_handle;
-    uint64_t last_valid_tail = first_blob_handle.offset;
-    uint64_t cur_valid_head = 0;
-    uint64_t last_discadable_tail = 0;
-
-    // TODO uint64_t before_size = record_iter->GetSize();
-    for (; record_iter->Valid(); record_iter->Next()) {
-      if (IsShutingDown()) {
-        s = Status::ShutdownInProgress();
-        return s;
+Status BlobGCJob::DigHoleFile(BlobFileMeta *input) {
+  Status s;
+  //open file
+  std::unique_ptr<PosixRandomRWFile> file;// TODO(@DorianZheng) set read ahead size
+  s = OpenBlobFile(input->file_number(), 0, db_options_, env_options_, env_, &file);
+  if (!s.ok()) {
+    return s;
+  }
+  auto record_iter = new BlobFileIterator(
+      std::move(file), input->file_number(), input->file_size(),
+      blob_gc_->titan_cf_options());
+  //pre size
+  uint64_t before_size = 0, after_size = 0;
+  record_iter->GetFileRealSize(&before_size);
+  //for each block-clusters
+  for (record_iter->SeekToFirst(); record_iter->Valid();) {
+    if (IsShutingDown()) {
+      s = Status::ShutdownInProgress();
+      return s;
+    }
+    //read first record in block-cluster and get hole size
+    BlobIndex blob_index = record_iter->GetBlobIndex();
+    BlobHandle blob_handle = blob_index.blob_handle;
+    uint64_t record_end = blob_handle.offset + blob_handle.size;
+    const uint64_t hole_start = blob_handle.offset;
+    const uint64_t hole_end = (record_end + block_size_) / block_size_ * block_size_;
+    bool enableHole = true;
+
+    while (record_end <= hole_end) {
+      //judge
+      if (enableHole) {
+        bool discardable = false;
+        s = DiscardEntry(record_iter->key(), blob_index, &discardable);
+        if (!s.ok()) {
```

Comment:
Why not return here?

Reply:
Because we need to ensure that the other records of the block-cluster are looped, and the next block-cluster is 4k aligned. The interface does not have the method of returning the corresponding iterator with offset, so we cannot jump directly to the next 4k aligned record. .